src/Sarif/Visitors/VersionOneLogicalLocationKeyToFullyQualifiedNameMappingVisitor.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT        
+// license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis.Sarif.VersionOne;
+
+namespace Microsoft.CodeAnalysis.Sarif.Visitors
+{
+    /// <summary>
+    /// This class visits all locations in a SARIF v1 log file. On observing use of result.logicalLocationKey,
+    /// which is used to disambiguate logical locations in the result.logicalLocations dictionary that have a 
+    /// common fully qualified name (but which are different types), the visitor creates a mapping between
+    /// the logical location key and its associated fully qualified name. This allows the v2 transformation
+    /// in particular to more easily populate its logical location equivalents. Additionally, the visitor 
+    /// stores a mapping from fully qualified name to decorated name, if one exists. These data (the
+    /// decorated name and fully qualified name) moved from the location object to the logical location
+    /// object in v2.
+    /// </summary>
+    public class VersionOneLogicalLocationKeyToFullyQualifiedNameMappingVisitor : SarifRewritingVisitorVersionOne
+    {
+        public VersionOneLogicalLocationKeyToFullyQualifiedNameMappingVisitor()
+        {
+            FullyQualifiedNameToDecoratedNameMap = new Dictionary<string, string>();
+            LogicalLocationKeyToFullyQualifiedNameMap = new Dictionary<string, string>();
+        }
+        public IDictionary<string, string> FullyQualifiedNameToDecoratedNameMap { get; set; }
+
+        public IDictionary<string, string> LogicalLocationKeyToFullyQualifiedNameMap { get; set; }
+
+        public override LocationVersionOne VisitLocationVersionOne(LocationVersionOne node)
+        {
+            if (!string.IsNullOrEmpty(node.LogicalLocationKey) &&
+                !string.IsNullOrEmpty(node.FullyQualifiedLogicalName) &&
+                !node.FullyQualifiedLogicalName.Equals(node.LogicalLocationKey))
```

Comment:
>( [](start = 54, length = 1)

`StringComparison.Ordinal`.

Reply:
This comparison (string.Equals) is one of the places in .NET where Ordinal is the default. This is true for other primitives such as the numerics. I don't make this explicit in code (for reasons of readability and maintainability) because I consider it a sort of essential fact about .NET that is helpful to know. (just as I suggest to people that they avoid unnecessary initialization of instance members). YMMV. :)

[https://docs.microsoft.com/en-us/dotnet/csharp/how-to/compare-strings](https://docs.microsoft.com/en-us/dotnet/csharp/how-to/compare-strings)
"The most common methods that test for equality, String.Equals and String.Equality, use a case-sensitive ordinal comparison."

---
In reply to: [244240657](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244240657) [](ancestors = 244240657)

Reply:
[I was not aware of that.](https://www.youtube.com/watch?v=Cd7tEsAuspA)

---
In reply to: [244376530](https://github.com/Microsoft/sarif-sdk/pull/1185#discussion_r244376530) [](ancestors = 244376530,244240657)